ntoskrnl/io/iomgr/deviface.c

```diff
@@ -412,6 +631,183 @@ IopOpenInterfaceKey(IN CONST GUID *InterfaceClassGuid,
+ * @implemented
+ *
+ * Returns the alias device interface of the specified device interface
+ * instance, if the alias exists.
+ * Documented in WDK.
+ *
+ * @param SymbolicLinkName
+ *        Pointer to a string which identifies the device interface instance
+ *
+ * @param AliasInterfaceClassGuid
+ *        See WDK
+ *
+ * @param AliasSymbolicLinkName
+ *        See WDK
+ *
+ * @return Three different NTSTATUS values in case of errors, and STATUS_SUCCESS
+ *         otherwise (see WDK for details)
+ *
+ * @remarks Must be called at IRQL = PASSIVE_LEVEL in the context of a system thread
+ *
+ *--*/
+NTSTATUS
+NTAPI
+IoGetDeviceInterfaceAlias(
+    _In_ PUNICODE_STRING SymbolicLinkName,
+    _In_ CONST GUID *AliasInterfaceClassGuid,
+    _Out_ PUNICODE_STRING AliasSymbolicLinkName)
+{
+    UNICODE_STRING AliasSymbolicLink = {0, 0, NULL};
+    UNICODE_STRING PrefixString;
+    UNICODE_STRING AliasGuidString;
+    UNICODE_STRING DeviceString, ReferenceString;
+    PKEY_VALUE_FULL_INFORMATION kvInfo;
+    HANDLE DeviceKey, AliasInstanceKey;
+    PVOID Buffer;
+    USHORT i;
+    NTSTATUS Status;
+
+    RtlInitEmptyUnicodeString(&PrefixString, NULL, 0);
+    RtlInitEmptyUnicodeString(&DeviceString, NULL, 0);
+    RtlInitEmptyUnicodeString(&ReferenceString, NULL, 0);
+    RtlInitEmptyUnicodeString(&AliasGuidString, NULL, 0);
+
+    DPRINT("IoGetDeviceInterfaceAlias() : symbolic link '%wZ'\n", SymbolicLinkName);
+
+    /* Sanity check */
+    if (SymbolicLinkName == NULL || AliasInterfaceClassGuid == NULL)
+    {
+        DPRINT1("IoGetDeviceInterfaceAlias() : invalid symbolic link or alias class GUID\n");
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    /* Convert alias GUID to a string */
+    Status = RtlStringFromGUID(AliasInterfaceClassGuid, &AliasGuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlStringFromGUID() failed with status 0x%08lx\n", Status);
+        goto Quit;
+    }
+
+    DPRINT("Alias GUID is '%wZ'\n", &AliasGuidString);
+
+    /* Get the device instance string of existing symbolic link */
+    Status = OpenRegistryHandlesFromSymbolicLink(SymbolicLinkName,
+                                                 KEY_QUERY_VALUE,
+                                                 NULL,
+                                                 &DeviceKey,
+                                                 NULL);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to open device instance key for '%wZ' with status 0x%08lx\n", SymbolicLinkName, Status);
+        goto Quit;
+    }
+
+    Status = IopGetRegistryValue(DeviceKey, L"DeviceInstance", &kvInfo);
+    ZwClose(DeviceKey);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed get device instance value with status 0x%08lx\n", Status);
+        goto Quit;
+    }
+
+    if (kvInfo->Type != REG_SZ || kvInfo->DataLength == 0)
+    {
+        DPRINT1("Wrong or empty instance value\n");
+        Status = STATUS_INSUFFICIENT_RESOURCES;
```

Comment:
An invalid registry value (or said registry value with no data on it) doesn't mean the system is under low system resources. It just means such registry data is corrupt or whatever. Perhaps using a status code like `STATUS_INVALID_PARAMETER` would be more appropriate as the caller passed a symbolic link name with invalid stuff.

Reply:
Fixed. Thank you. :slightly_smiling_face: